### ISSUE TITLE
POSIX shared memory APIs

### ISSUE THREAD

**Issue by karthikshanmugam** [GENIVI/dlt-daemon#90]
The DLT implementation uses non-POSIX shared memory and semaphore API like shmget, semget etc.,
These APIs are not availble in POSIX, the suggestion is to use shm_open, sem_open APIs. I can submit a patch if needed.

**Comment by visglz** [comment]
Is the patch already available somewhere, maybe as PR? I would also be interested in compiling DLT for non linux POSIX platforms.

**Comment by clipka** [comment]
No, we do not have it in place for shared memory. You can compile it with communication over sockets for QNX by using current master.

**Comment by ssugiura** [comment]
@karthikshanmugam Sorry that we haven't updated for long time. I want to make sure your expectation here. We do have an option for shared memory as IPC, but actually it depends on FIFO or unix socket to send the message header. So even shared memory is enabled, FIFO or unix socket is used, which won't solve the situation you mentioned https://github.com/GENIVI/dlt-daemon/issues/52. We do so because dlt-daemon needs to wait for both library and client, which for client only socket connection is available.
So the question is, is it OK to use FIFO/unix socket+shared memory, or you want the complete shared memory IPC?

**Comment by karthikshanmugam** [comment]
Hello,

Sorry for the delayed response.

Having the header exchanged via socket will still not solve the problem.

Would POSIX message queue be an option instead of sockets to exchange
header and ShMem for data?

Best regards,
Karthik

**Comment by ssugiura** [comment]
@karthikshanmugam I guess it's the same for message queue. We need to create a queue beforehand on QNX, which I think is no difference to the network stack problem.
So the option we have is, either getting another thread in dlt-daemon to wait on library message, or retrying socket creation in library (easier, but should not be part of upstreamed code in my feeling).

**Comment by karthikshanmugam** [comment]
It is not same, the problem with the socket is that we cannot have two
network interfaces for an application. Message queues does not have this
limitation. It is okay to transfer header over message queue (instead of
sockets) and data over shared memory

**Comment by ssugiura** [comment]
@karthikshanmugam Could you give me more details why you cannot have two network interfaces for an application? Also, I would like to know how you start network stacks mentioned in https://github.com/GENIVI/dlt-daemon/issues/52.
> Default Network Stack -> UDS communication between processess, launched earlier in the boot.
Debug Network Stack -> USB to Ethernet communication with debug IP address, launched later in the boot.
Vehicle Network Stack -> OABR Ethernet with vehicel IP address, launched upon request.

**Comment by ssugiura** [comment]
@karthikshanmugam Do you have any update on this?

**Comment by karthikshanmugam** [comment]
Hello Saya,

Thank you for reminding me. I somehow missed the earlier email.

In QNX, we use the io-pkt-* services to launch the network stacks.
Even though we can use mount command to mount the network stack in the same
namespace for security purposes we use different network stack.

Best regards,
Karthik

**Comment by ssugiura** [comment]
@karthikshanmugam Thank you for your reply. I understood your use case.
Honestly, it would be a bit difficult as it's a big change to dlt-daemon in order to do so. dlt-daemon needs to wait for file descriptor from both library (FIFO or unix socket) or client (TCP/IP or serial), so to support shared memory or message queues, we need to wait on e.g. different thread.

If you already have any ideas, could you please give us the commits? If not, I'm afraid we cannot do the fix soon. On the other hand, replacing non-POSIX API to POSIX ones from the original issue of this thread, we can merge the fix soon.

**Comment by karthikshanmugam** [comment]
Hello Saya,

I do not have any patches now. Please go ahead and replace the non-POSIX to
POSIX APIs that would be of tremendous help.

Best regards,
Karthik

**Comment by ssugiura** [comment]
@karthikshanmugam OK, we raise a pull request to replace non POSIX APIs to POSIX API. https://github.com/GENIVI/dlt-daemon/pull/151
Please check it.

**Comment by ssugiura** [comment]
@karthikshanmugam I pushed the changes. Please refer to the latest commit. I will close this issue as of now. If further work is needed, we can re-open or create a new issue.

**Comment by karthikshanmugam** [comment]
Thank you. I will check and let you know.

Best regards,
Karthik
